### ISSUE TITLE
Search & Modify with Buffer object

### ISSUE THREAD

**Issue by lennarkivistik** [ldapts/ldapts#11]
Hi,
Superb library!

Is it possible to add a ignore parsing array option on client.modify and client.search to receive the raw buffer instead of the parsed response, ( in SearchEntry.js ) as it converts everything to a string.

`{
          scope: 'sub',
          filter: '(&(sAMAccountName=someAccount))',
          attributes: ['thumbnailPhoto', 'dn', 'name', 'something...'],
          noparsing: ['thumbnailPhoto'],
}`

essentially im trying to get the thumbnailPhoto which is usually a binary photo
and then save the photo to file, but as it is already converted to string, some characters gets mangled by the encoding and Im unable to get the correct output.

ldapjs lib has the same issue where object.entries are strings but since it pushes the full object back, the data was still accessable at object.attributes[0].buffers[0];

`Buffer.from(object.attributes[0].buffers[0], 'hex').toString('binary');`

not totally sure how to send a "modify" replace in binary form, clues?

**Comment by jgeurts** [comment]
I've added support for binary attribute values. If the [binary subtype is specified](https://docs.oracle.com/cd/E19424-01/820-4809/bcacz/index.html), then the attribute value will be a Buffer array. Otherwise, it will be a string array.

To update the thumbnailPhoto with a new picture, you should be able to do the following:
```
const thumbnailPhotoBuffer = await fs.readFile(path.join(__dirname, './groot_100.jpg'));
await client.modify('uid=groot,ou=Users,dc=foo,dc=com', new Change({
  operation: 'replace',
  modification: new Attribute({
    type: 'thumbnailPhoto;binary',
    values: [thumbnailPhotoBuffer]
  }),
}));
```

Hope that helps!

**Comment by jgeurts** [comment]
Likewise, if the server sends a binary type, the attribute value will be a Buffer array instead of a string array. That should take care of search results. Please let me know if you have troubles with this - I don't have access to an ldap server that has a binary attribute, unfortunately.

**Comment by lennarkivistik** [comment]
**Kudos!**
Works as described, subtype worked nicely for reading binary response
as did the example code you specified for writing to ldap!

Thanks for the help! @jgeurts

**Comment by jgeurts** [comment]
Awesome, glad that worked!

**Comment by RedShift1** [comment]
This doesn't seem to work with Active Directory, where for example it's important to keep the binary version of objectGuid. Output from a search operation:

```
{ searchEntries:
   [ { objectGUID: '7\t��\u0013p;C������>�' },
     { objectGUID: '�\tk7���A��Ĉ�R\u0015o' } ],
  searchReferences: [] }
```

So we may want to have the option to retrieve the original buffer, in case the server doesn't report it as a binary field but you want to access it like that anyway.

**Comment by lennarkivistik** [comment]
It works for me, 

so for example doing a search like this with ldapts towards your Active Directory server will give me a list of all objectGUIDs for users
```
        const entries = await this.client.search(LDAP_USERS_OU, {
          scope: 'sub',
          filter: LDAP_TREE_FILTER,
          attributes: [
            'objectGUID;binary'
          ],
          sizeLimit: 500,
          paged: true
        });
```

Then converting it to which ever format I want afterwords like this:
```
entries.map(e => Buffer.from(e['objectGUID;binary'], 'binary').toString('hex'))
```

In this instance the hex matches exactly the hex in Active Directory on a specific user
if you go to the user properties --> attribute editor --> find objectGUID --> look at the hex code.

**Comment by jgeurts** [comment]
@RedShift1 please let me know if the example that @lennarkivistik wrote works for you. Mainly appending `;binary` to the attribute name. 

If it doesn’t work, I’ll likely need a test that reproduces what you’re seeing in order to fix things.

**Comment by RedShift1** [comment]
Sorry I was mistaken, it didn't register with me I had to append ";binary" to the attribute name, I thought it _the server that was sending data with a binary attribute_ and that would automatically be picked up by the library which would then pass the buffer instead of a string. Never mind me then...

**Comment by kabo** [comment]
Appending `;binary` doesn't seem to work on AWS SimpleAD.

```typescript
    client.search(adBase, {
      filter: `(cn=someGroup)`,
      scope: 'sub',
      attributes: [ 'objectSid', 'objectSid;binary' ],
    })
      .then(console.log)
```
prints

```typescript
{ searchEntries:
   [ { dn:
        'CN=someGroup,CN=Users,DC=xxx,DC=yyy,DC=zzz',
       objectSid:
        '\u0001\u0005\u0000\u0000\u0000\u0000\u0000\u0005\u0015\u0000\u0000\u0000����%(�Y�}\\��\b\u0000\u0000',
       'objectSid;binary': [] } ],
  searchReferences: [] }
```

**Comment by jgeurts** [comment]
I can take a look at what ldapjs is doing with the `raw` property, but I won't be able to test very well it as I don't have a reproducible ldap server for this issue.

**Comment by Toxicable** [comment]
For some reason something transforms our input attribute `'thumbnailPhoto;binary'` into `THUMBNAILPHOTO;BINARY`

So I bad to modify the lib to do 
```
    _isBinaryType() {
        return this.type.endsWith(';binary') || this.type.endsWith(';BINARY')
    }
```
Once I did that it worked for me when doing a search, the result was already a `Buffer` so I didn't have to do what @lennarkivistik suggested.

**Comment by jgeurts** [comment]
@Toxicable Thanks for letting me know! Release 2.0.2 should have this fixed for you!

**Comment by BytewaveMLP** [comment]
I'm having the same issue as @kabo when attempting to retrieve the `jpegPhoto` attribute from users via a search. Attempting to retrieve `jpegPhoto` returns a corrupted binary string, but attempting to search for the attribute `jpegPhoto;binary` returns an empty array.

Here is a minimal example:

```ts
const searchResults = await ldapClient.search(CONFIG.get('ldap.userDN'), {
	filter: searchFilter,
	attributes: [
		'jpegPhoto',
		'jpegPhoto;binary',
	],
});
console.log(searchResults.searchEntries);
```

```js
{
	dn: 'uid=bytewavemlp,dc=redacted',
	jpegPhoto: '����\u0000...',
	'jpegPhoto;binary': []
}
```

Using OpenLDAP 2.4.50 (via Docker image `osixia/openldap:1.3.0`) and `ldapts` 2.5.1.

**Comment by BytewaveMLP** [comment]
Upon further investigation, it seems like this is just an issue with the LDAP standard itself (and maybe OpenLDAP adhering too strongly to the standard as well?). I did a lot of digging, and discovered that the binary attribute option (`;binary`) only applies to certain attributes types ([RFC4522 section 3](https://tools.ietf.org/html/rfc4522#section-3)):

> An attribute description SHALL be treated as unrecognized if it contains the binary option and the syntax of the attribute does not have an associated ASN.1 type [RFC4517], or the BER encoding of values of that type is not supported.

I'm not entirely sure, but it seems like the syntax type `jpegPhoto` uses in OpenLDAP (`1.3.6.1.4.1.1466.115.121.1.28` "JPEG") is one of those types that, for whatever reason, doesn't support BER encoding? So, OpenLDAP returns no value when requesting `jpegPhoto;binary` in a search because it's treating it as unrecognized, I guess? I **was** able to retrieve an X.509 certificate for a user with the `userCertificate` attribute type in `binary` transfer mode, as that's one of the attribute types the RFC focuses on supporting in the first place, so I can confirm that OpenLDAP *does* support `;binary`, just not for all attributes.

It would be nice to have a raw mode, or access to the raw data returned by the LDAP server before it gets `toString`ed (perhaps in a separate `SearchResults` key?). I'll open a new issue to track this request and link it here for reference.

**Comment by peleglahav** [comment]
TL;DR - decoding issues still appearing when retrieving `objectSid` and `objectGUID`.

I'm trying to run the following:
    ```
const resp = await client.search(searchDN, {
        filter: '(objectClass=user)',
        explicitBufferAttributes: ['objectSid'],
      });

      console.log(resp);
      console.log(resp.searchEntries[0].objectSid);
```

`resp.searchEntries[0].objectSid` indeed returns as a buffer, but when I run:
`resp.searchEntries[0].objectSid.toString('hex') `
I'm not getting the correct value as I expect something of the format S-1-234...

Any idea how to solve?

**Comment by ayZagen** [comment]
> This is not related to this library - the binary data you get from objectSid is correct
@RedShift1 that's right.

here is an example implementation [objectsid.js](https://github.com/pfmooney/node-ldapjs-mangle-proxy/blob/master/lib/objectsid.js)
